draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -2204,68 +2249,81 @@ Input: u, an element of F.
 Output: (x, y), a point on E.
 
 Constants:
-1. c1 = (2 * q - 1) / 3   // Integer arithmetic
+1. c1 = (2 * q - 1) / 3     # Integer arithmetic
 
 Steps:
-1. t1 = u^2
-2. t1 = t1 - B
-3.  x = t1^c1             // x = (u^2 - B)^((2 * q - 1) / 3)
-4.  y = u
+1. tv1 = u^2
+2. tv1 = tv1 - B
+3.   x = tv1^c1             # x = (u^2 - B)^((2 * q - 1) / 3)
+4.   y = u
 5. return (x, y)
 ~~~
 
-### Elligator 2, A == 0 Method {#ell2a0}
+### Elligator 2, C == 0 Method {#ell2c0}
 
-The function map\_to\_curve\_ell2A0(u) implements an adaptation of Elligator 2
-{{BLMP19}} targeting curves given by y^2 = x^3 + B * x over F such that q = 3 (mod 4).
+The function map\_to\_curve\_ell2C0(u) implements an adaptation of Elligator 2
+{{BLMP19}} targeting curves given by the Weierstrass equation
 
-Preconditions: An elliptic curve over F such that q = 3 (mod 4).
+~~~
+    Y^2 = X^3 + D * X
+~~~
 
-Constants: B, the parameter of the elliptic curve.
+over F = GF(q), q = 3 (mod 4).
+This method also works for curves given by the Montgomery equation
 
-Sign of y: Inputs u and -u give the same x-coordinate.
-Thus, we set sgn0(y) == sgn0(u).
+~~~
+    K * t^2 = s^3 + s
+~~~
+
+via the change of variables given in {{montgomery}}.
+
+Preconditions: A curve whose equation and base field meet the requirements immediately above.
+
+Constants: D, the parameter of the elliptic curve.
+
+Sign of Y: Inputs u and -u give the same X-coordinate.
+Thus, we set sgn0(Y) == sgn0(u).
 
 Exceptions: none.
 
 Operations:
 
 ~~~
-1.  x1 = u
-2. gx1 = x1^3 + B * x1
-3.  x2 = -x1
-4. gx2 = -gx1
-5. If is_square(gx1), set x = x1 and y = sqrt(gx1)
-6. Else set x = x2 and y = sqrt(gx2)
-7. If sgn0(u) != sgn0(y), set y = -y.
-8. return (x, y)
+1.  X1 = u
+2. gX1 = X1^3 + D * X1
+3.  X2 = -X1
+4. gX2 = -gX1
+5. If is_square(gX1), set X = X1 and Y = sqrt(gX1)
+6. Else set X = X2 and Y = sqrt(gX2)
+7. If sgn0(u) != sgn0(Y), set Y = -Y
+8. return (X, Y)
 ~~~
 
 #### Implementation
 
-The following procedure implements the Elligator 2 mapping for A == 0
+The following procedure implements the Elligator 2 mapping for C == 0
 in a straight-line fashion.
 
 ~~~
-map_to_curve_ell2A0(u)
+map_to_curve_ell2C0(u)
 Input: u, an element of F.
-Output: (x, y), a point on E.
+Output: (X, Y), a point on E.
 
 Constants:
-1. c1 = (p + 1) / 4         // Integer arithmetic
+1. c1 = (q + 1) / 4         # Integer arithmetic
```

Comment:
This constant can be omitted since calling to sqrt already includes this constant as q=3 (mod 4).
Replace 
```
6.   Y = gX1^c1
```
by 
```
6.   Y = sqrt(gX1)
```

Reply:
This would be a nice simplification, but I don't think it works in this context.

The reason is, gX1 is not guaranteed to be square, and our sqrt functions are explicitly only defined for square inputs. The point here is that we're *not necessarily* taking the sqrt of gX1---we might instead get back the sqrt of -gX1.

Reply:
oh, I got it.
So using the exponentiation you are preventing (for instance) the following case: someone implemeted sqrt in such a way that always returns error when the sqrt is not found.

Reply:
Yes, exactly right. Nicely put.